app/services/developers/event_publisher.rb

```diff
@@ -0,0 +1,7 @@
+module Developers
+  class EventPublisher
+    def execute(context, event_type, actor, resource)
+      ActiveSupport::Notifications.instrument("#{event_type}.pupilfirst", resource_id: resource.id, actor_id: actor.id, context_id: context.id)
```

Comment:
It would probably be best not to pass `context` / `course` via instrumentation. This was used by the webhook delivery class only to fetch the URL endpoint.

We'll also plan to change the storage of webhook endpoints to `school` later on.

Reply:
> It would probably be best not to pass context / course via instrumentation. This was used by the webhook delivery class only to fetch the URL endpoint.

It is fine for me. I do not use it now, but I've implemented only 3 subscribers for now, I think I could need it sooner or later. But maybe we should follow YAGNI :) 

> We'll also plan to change the storage of webhook endpoints to school later on.

That's why I've called in `context` not the `course`. School is also some context for published events.

Reply:
@mpraglowski What purpose is `context` / `course` serving here? The event's `resource` is being passed anyway, along with the `actor` / `user`.

Strictly speaking, only the `resource` is required, since it would be possible to load the `actor` and the related `context` / `course` through the `resource`.

I understand that having `actor_id` / `user_id` sent via the instrumentation would be convenient in that it could help avoid one joined query, but the need for `context` / `course` escapes me.

This reply also applies to [your comment below](https://github.com/pupilfirst/pupilfirst/pull/629/files#r576614208).

Reply:
Ok, removal of `context` / `course` forces this code to be aware of type of resource passed. In current solution when `context` / `course` comes as method argument it is a generic thing and any class that implements `webhook_endpoint` is fine here (currently only `course` so it is ok to rename it).

```ruby
module WebhookDeliveries
  class CreateService
    def execute(context, event_type, resource)
      webhook_endpoint = context.webhook_endpoint
      return unless webhook_endpoint&.active?
      return unless event_type.in? webhook_endpoint.events
      WebhookDeliveries::DeliverJob.perform_later(event_type, context, resource)
    end
  end
end
```

Other way to ensure this works fine is that any `resource` passed as argument here must have `course` getter available. But I will soon (end of week) push a PR with more places where `Developers::NotificationService` will be used and in some cases I think `resource` will be the `course` itself.  I would keep current solution with `context` passed as argument - it gives more flexibility here - WDYT?